### PR TITLE
R2LPL two-sided teacher: release bands, take-off pools, trainer override channel

### PR DESCRIPTION
**Pull request by danielsanchezaran**
## What

Bakes the "two-sided teacher" data treatment into the R2LPL lifelong workflow, and restores wrapper-driven training on tier4-main:

1. **`rlvr/autoresearch/tools/build_release_bands.py`** (new): for every mined event, extract recorded **post-offense release-band** frames from the source sequence via chunk-manifest provenance, kept by traffic-light alignment — route-lane RED (and not also green) frames teach patience unconditionally; GREEN/no-TL frames are kept only when the recorded future genuinely departs (threshold required — creeps are rejected on purpose). Frames are indexed by (directory, filename-prefix) log identity and the band walk follows the frames that actually exist, so step>1 catalogs work.
2. **`release_bands` workflow block** in `run_lifelong_r2lpl_rounds`: per round, extract bands from the round's own `credit_windows.jsonl`, sample them at a configured ratio to the focus count, and union them into the training list. All knobs required, validated at startup (ranges, manifest existence, base_sft-only); the workflow contract carries the block through (empty block rejected loudly).
3. **`build_r2lpl_pools takeoff`** (new stage): stopped-input + route-lane-green + genuinely-departing-future pool; stratified into the anchor slice via `training.anchor.takeoff_scene_list`/`takeoff_fraction` (validated with the waits stratum: paired keys, (0,1) range, fraction sum < 1, list existence — all at startup).
4. **`TrainConfig.train_overrides_json`** (the only `diffusion_planner/` change, +36 lines, additive): the CLI consolidation left most TrainConfig fields without command-line handles, but launch wrappers must still SET them per run — a warm-started fine-tune silently training at the from-scratch defaults (e.g. `learning_rate` 1e-4 instead of an intended 5e-6) is the failure this prevents. One `cli()` field pointing at a JSON of overrides; unknown keys, CLI-shadowing keys, and scalar type mismatches (`"false"` as a truthy string bool, float-for-int) all fail loudly. The round runner emits `cli()`-exposed knobs as flags (introspected, not hardcoded) and everything else through `train_overrides.json` written next to the round's checkpoints; `train_args` attempting to set runner-owned keys (`train_epochs`, scene lists, resume checkpoint) is rejected instead of silently overriding computed values.

## Why

Mining carves credit windows that end at the offense, so every repaired target teaches the entry into a braking manoeuvre and never its resolution. A corpus of one-sided windows is systematically deceleration-flavoured, and each training epoch absorbs more of the bias — observed as a monotone per-epoch erosion of take-off response at heavily-mined geometries while stop compliance persists. A controlled run with the repaired slice removed (all else identical) eliminated the erosion, isolating the cause to window geometry. Pairing each event's window with frames from the same event's recorded post-offense band cancels the bias at the source: extended training then holds/improves the take-off response instead of eroding it, while retaining the repaired corpus's reward value. Nothing is generated, relabeled, or filtered out of the repaired set.

## Impact on existing users

- **IL training: byte-identical.** `train_overrides_json` defaults to `""` → strict no-op; the `train_run.py` relaunch path (`to_command_line`) omits default-valued flags, so existing command lines/logs don't change. Verified by running the IL parse→build→relaunch path with no new flag.
- **R2LPL configs without `release_bands`/`takeoff_*`: unchanged** (blocks are opt-in; absent = disabled, empty = loud error).
- Note: without commit 4, `run_lifelong_r2lpl_rounds` base_sft training cannot launch against this repo's trainer at all (the runner passes flags the consolidated CLI no longer accepts) — this PR is also the fix for that.

## Validation

- 178 tests green (`test_r2lpl_lifelong_replay.py`), including: extractor TL/creep alignment on synthetic sequences, unmatched-chunk-key and zero-row loud failures, contract carriage, anchor stratum validation (pairing/range/sum), ratio-sampling + seed-offset + rewrite wiring, reserved-key rejection, override type/CLI-shadow rejections.
- Extraction equivalence: the tool reproduces a previously hand-built release-band set **exactly** (set-identical rows) from the same credit windows + manifest.
- End-to-end 1-round workflow smoke (mine → repair → auto-extract → sample → union → 1-epoch base_sft train → checkpoint) passed with the block enabled.
- Cyclomatic complexity audit (radon): all touched functions ≤ C; the two runner functions this PR grew were refactored back to/below their previous grades (`_base_train_invocation` C14→C12, `_anchor_slice_paths` C16→C19 with the new stratum, helpers A/B).
- Two independent review passes; all findings addressed in dedicated commits.

### CODE REVIEW

**Review by danielsanchezaran — COMMENTED** [review]
I found three selection-semantics issues that should be fixed before merge. The main one affects both new teacher pools: zero-padded future rows can turn a short valid motion into a long path by adding the jump back to the origin. The anchor mixer can also eliminate the normal stratum despite accepting a fraction sum below 1, and a positive but sub-frame release window silently degenerates to the offense frame only.

I ran the full focused suite on this exact head: 177 passed, 1 skipped. Ruff and Python compilation also pass for every changed Python file. There are no GitHub checks attached to the head. GitHub does not allow me to submit Request changes on my own PR, so I am leaving these as blocking inline comments instead.

File: rlvr/autoresearch/tools/build_release_bands.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        m = _FRAME_RE.match(f)
+        if m and m.group(1) == prefix:
+            idx[int(m.group(2))] = os.path.join(d, f)
+    if not idx:
+        raise ValueError(f"no frames of log {prefix!r} next to {scene_path}")
+    return idx
+
+
+def _pathlen(xy: np.ndarray) -> float:
+    return float(np.linalg.norm(np.diff(xy, axis=0), axis=1).sum())
```
Comment (danielsanchezaran) [inline rlvr/autoresearch/tools/build_release_bands.py:105]: This includes invalid zero-padded future rows in the path length. I reproduced a green frame with two valid points 1 m apart followed by zero padding: this function reports 7 m because it counts the jump from the last valid point back to `(0, 0)`, so a 5 m threshold incorrectly keeps it. This repository explicitly documents `ego_agent_future` as zero-padded on invalid steps (for example in `add_distractor_neighbors_npz.py`). Please mask invalid future rows using the canonical validity rule before taking the diff, and add a padded-future regression test; otherwise the release corpus admits short/truncated trajectories as genuine departures.

File: rlvr/autoresearch/tools/build_r2lpl_pools.py [hunk 2]
```diff
@@ -134,6 +136,63 @@ def _moving_filter(paths: list[str], thresh: float, workers: int) -> tuple[list[
...
+            recent = float(
+                np.linalg.norm(np.diff(past[-(recent_steps + 1) :], axis=0), axis=1).sum()
+            )
+            if recent >= max_recent_travel_m:
+                return (path, False)
+            green, _red = _route_tl_state(d)
+            if not green:
+                return (path, False)
+            fut = d["ego_agent_future"][:, :2]
+            travel = float(np.linalg.norm(np.diff(fut, axis=0), axis=1).sum())
```
Comment (danielsanchezaran) [inline rlvr/autoresearch/tools/build_r2lpl_pools.py]: The zero-padding path-length bug is duplicated in the take-off filter rather than going through a shared valid-path helper. With valid future points at x=5 and x=6 followed by zeros, `_takeoff_worker(..., min_future_travel_m=5)` returns `True` even though valid travel is only 1 m; the computed 7 m includes the jump back to the padded origin. Please apply the future validity mask here too (preferably via one shared helper) and cover this worker with the same padded-future case.

File: rlvr/autoresearch/tools/run_lifelong_r2lpl_rounds.py [hunk 3]
```diff
@@ -1220,15 +1259,10 @@ def _sample(pool: list[str], n: int) -> list[str]:
...
-        frac = float(waits_fraction)
-        if not 0.0 < frac < 1.0:
-            raise ValueError(f"training.anchor.waits_fraction must be in (0, 1): {frac}")
-        waits_pool = [str(p) for p in _read_json_list(Path(waits_list))]
-        if not waits_pool:
-            raise ValueError(f"training.anchor.waits_scene_list is empty: {waits_list}")
-        n_waits = int(round(frac * n_anchor))
-        picked.extend(_sample(waits_pool, n_waits))
+    for pool, frac in strata:
+        n_slice = int(round(frac * n_anchor))
```
Comment (danielsanchezaran) [inline rlvr/autoresearch/tools/run_lifelong_r2lpl_rounds.py]: Rounding each stratum independently can consume the entire anchor budget even though validation promises to leave room for the normal slice. For example, `waits_fraction=0.4`, `takeoff_fraction=0.4`, and `n_anchor=2` is accepted because the fractions sum to 0.8, but both quotas round to 1 and the result contains zero normal anchors. Please allocate the integer quotas jointly (or floor the special strata and explicitly reserve the configured normal remainder), and add a small-budget test so the validated composition is preserved.

File: rlvr/autoresearch/tools/build_release_bands.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    manifest = _load_chunk_manifest(chunk_manifest)
+    events = _load_event_offenses(credit_windows)
+    unmatched = sorted(k for k in events if k not in manifest)
+    if unmatched:
+        raise ValueError(
+            f"{len(unmatched)}/{len(events)} event chunk keys are absent from "
+            f"{chunk_manifest} (first: {unmatched[0]!r}) — wrong manifest for this run?"
+        )
+
+    post_frames = int(round(post_window_s * frame_hz))
```
Comment (danielsanchezaran) [inline rlvr/autoresearch/tools/build_release_bands.py]: A positive duration does not guarantee a positive post-offense window after this conversion. At 10 Hz, `post_window_s=0.01` passes both validators but rounds to zero; on a red offense fixture the extractor succeeds and emits only the offense frame, defeating the advertised post-offense/two-sided corpus semantics. Please reject configurations where the duration converts to fewer than one frame (in both startup validation and this standalone builder), or use an explicitly documented ceiling policy.

**Review by danielsanchezaran — COMMENTED** [review]
Follow-up on the current head: it is byte-for-byte unchanged from yesterday's reviewed commit, and all three blockers still reproduce.

- A future with 1 m of valid motion followed by zero padding is still measured as 7 m, and the take-off worker accepts it at a 5 m threshold ([release-band thread](https://github.com/tier4/Diffusion-Planner/pull/379#discussion_r3879020180), [take-off thread](https://github.com/tier4/Diffusion-Planner/pull/379#discussion_r3879020186)).
- With two anchors and 0.4 waits / 0.4 take-off fractions, both special strata round to one and the accepted configuration still produces zero normal anchors ([anchor thread](https://github.com/tier4/Diffusion-Planner/pull/379#discussion_r3879020188)).
- At 10 Hz, a positive 0.01-second post window still succeeds and emits only the offense frame ([window thread](https://github.com/tier4/Diffusion-Planner/pull/379#discussion_r3879020193)).

I reran the focused suite today: 177 passed, 1 skipped. Ruff and Python compilation also pass. No GitHub checks or commit statuses are attached to this head. These issues should still be fixed before merge; I am not duplicating the existing inline comments.

File: rlvr/autoresearch/tools/build_release_bands.py (hunk 1 above)
Comment (danielsanchezaran) [reply rlvr/autoresearch/tools/build_release_bands.py:105]: Fixed in 91a29db7: extracted a shared `valid_future_pathlen` that masks zero-padded rows (|xy|.sum > 1e-6, same rule as `_valid_path_indices`) before the arc-length sum — your 1 m-valid/7 m-reported repro is now a regression test (`test_future_pathlen_masks_zero_padding`: a padded green pseudo-departure is dropped, only the red patience frame kept). Fix was authored Friday; GH013 blocked all weekend pushes (see follow-up review). Measured impact on the shipped ep85/ep90 corpora is zero — 0 padded futures / 0 decision flips in 3k band samples (the contiguous 0.1 s catalog stores full 80-step futures).

File: rlvr/autoresearch/tools/build_r2lpl_pools.py (hunk 2 above)
Comment (danielsanchezaran) [reply rlvr/autoresearch/tools/build_r2lpl_pools.py]: Fixed in 91a29db7: `_takeoff_worker` now imports the same `valid_future_pathlen` instead of its own raw diff — no duplicated logic. Your x=5,x=6-then-zeros case is covered in the regression test (padded creep rejected at a 4 m threshold). Zero impact on the shipped take-off pool: 0 padded futures / 0 flips in 2k samples.

File: rlvr/autoresearch/tools/run_lifelong_r2lpl_rounds.py (hunk 3 above)
Comment (danielsanchezaran) [reply rlvr/autoresearch/tools/run_lifelong_r2lpl_rounds.py]: Fixed in 91a29db7: stratum quotas are now floored jointly (`floor(frac * n_anchor)`), so with sum(fractions) < 1 the floors sum to <= n_anchor-1 and the normal slice always survives — your 2-anchor/0.4+0.4 case is the regression test (`test_anchor_slice_paths_small_budget_keeps_normal_slice`). Structurally irrelevant to the shipped corpora (large budgets), but correct now at any size.

File: rlvr/autoresearch/tools/build_release_bands.py (hunk 4 above)
Comment (danielsanchezaran) [reply rlvr/autoresearch/tools/build_release_bands.py]: Fixed in 91a29db7: `post_window_s * frame_hz` rounding to zero frames is now rejected in BOTH the band builder and the runner's startup validator (`_validate_release_bands_config`) — 0.01 s at 10 Hz raises at startup instead of silently emitting only the offense frame (`test_build_release_bands_rejects_subframe_window` covers both sites). Local validation for the whole commit: 181 tests passed, ruff + pre-commit clean.

**Review by danielsanchezaran — COMMENTED** [review]
Follow-up on `25cf2f052`: the original review blockers are resolved, and I found no new correctness issues in the fixes.

I reran the exact counterexamples independently: the padded future now measures 1 m and is rejected by the 5 m take-off threshold; the 0.4 + 0.4 small-budget anchor case preserves normal scenes; and the 0.01-second window at 10 Hz now fails loudly instead of emitting only the offense frame. The expanded focused suite passes with 180 passed and 1 environment-dependent GPU test skipped. Ruff, Python compilation, and `git diff --check` also pass.

There are no GitHub checks or commit statuses attached to this head. GitHub does not allow me to approve my own PR, so this is submitted as a comment review, but the concerns from my previous review are cleared.

**Review by danielsanchezaran — COMMENTED** [review]
Follow-up on `e6fb8651f`: the cyclomatic-complexity refactor is behavior-preserving in the reviewed paths, and I found no new correctness issues.

The full focused suite passes with 180 passed and 1 environment-dependent GPU test skipped. Ruff, Python compilation, and `git diff --check` pass. I also compared the pre-refactor and current release-band selection, padded-future path length, small-budget anchor composition, and sub-frame validation outputs directly; they match exactly. There are no GitHub checks or commit statuses attached to this head. GitHub does not allow me to approve my own PR, so this is submitted as a comment review, but the current head is clear from this review.

**Review by danielsanchezaran — COMMENTED** [review]
Follow-up on `00319b90f`: the new `ego-dims` platform slice reads the documented wheelbase field, preserves input order, reports unreadable scenes, and fails loudly when a non-empty input produces no usable output. I found no new correctness issues, and the earlier release-band, padding, anchor-allocation, and sub-frame-window fixes remain intact.

The full focused suite passes on both the exact head and GitHub's current test merge with 181 passed and one environment-dependent GPU test skipped. Ruff, formatting, Python compilation, and `git diff --check` pass. There are no GitHub checks attached to this head. GitHub does not allow me to approve my own PR, so this is submitted as a comment review, but the current head is clear from this review.

File: diffusion_planner/diffusion_planner/config/train_config.py [hunk 5]
```diff
@@ -46,6 +46,13 @@ class TrainConfig(ClosedLoopConfig, ScenarioOpenLoopConfig, ModelConfig):
     ego_past_noise_std: float = 0.1
     use_smoothing_future_trajectory: bool = True
     normalization_file_path: str = "normalization.json"
+    # Override channel for launch wrappers: JSON object of non-CLI TrainConfig
+    # fields; unknown keys fail loudly (see train_predictor.apply_overrides_json).
+    train_overrides_json: str = cli(
+        "JSON object of TrainConfig field overrides applied after CLI parsing",
+        path=True,
+        default="",
+    )
```
Comment (SakodaShintaro) [inline diffusion_planner/diffusion_planner/config/train_config.py:55]: I thought it might be clearer to wrap all the necessary arguments with `cli`, or perhaps adopt uv.

**Review by SakodaShintaro — APPROVED** [review]
